util/dvsim/SgeLauncher.py

```diff
@@ -0,0 +1,174 @@
+#      SgeLauncher Class
+#
+# ------------------------------------
+import os
+import shlex
+import subprocess
+import SGE
+from subprocess import PIPE, Popen
+from Launcher import ErrorMessage, Launcher, LauncherError
+global job_name
+
+pid = os.getpid()
+
+
+class SgeLauncher(Launcher):
+    """
+    Implementation of Launcher to launch jobs in the user's local workstation.
+    """
+
+    # Misc common SgeLauncher settings.
+    max_odirs = 5
+
+    def __init__(self, deploy):
+        '''Initialize common class members.'''
+
+        super().__init__(deploy)
+
+        # Popen object when launching the job.
+        self.process = None
+
+    def _do_launch(self):
+        global job_name
+        # Update the shell's env vars with self.exports. Values in exports must
+        # replace the values in the shell's env vars if the keys match.
+        exports = os.environ.copy()
+        if self.deploy.exports:
+            exports.update(self.deploy.exports)
+
+        # Clear the magic MAKEFLAGS variable from exports if necessary. This
+        # variable is used by recursive Make calls to pass variables from one
+        # level to the next. Here, self.cmd is a call to Make but it's
+        # logically a top-level invocation: we don't want to pollute the flow's
+        # Makefile with Make variables from any wrapper that called dvsim.
+        if 'MAKEFLAGS' in exports:
+            del exports['MAKEFLAGS']
+
+        self._dump_env_vars(exports)
+
+        try:
+            f = open(self.deploy.get_log_path(),
+                     "w",
+                     encoding="UTF-8",
+                     errors="surrogateescape")
+            f.write("[Executing]:\n{}\n\n".format(self.deploy.cmd))
+            f.flush()
+            # ---------- prepare SGE job struct -----
+            sgeJob = SGE.qsubOptions()
+            sgeJob.args.N = 'VCS_RUN_' + str(pid)  # Name of Grid Engine job
+            if "build.log" in self.deploy.get_log_path():
+                sgeJob.args.N = 'VCS_BUILD_' + str(pid)  # Name of Grid Engine job
+
+            job_name = sgeJob.args.N
+            sgeJob.args.t = '0'      # Define an array job with 20 subjobs
+            sgeJob.args.slot = '1'      # Define num of slot
+            sgeJob.args.sync = 'y'      # wait for the job to complete before exiting
+            sgeJob.args.q = 'vcs_q'   # Define the sge queue name
+            sgeJob.args.l = 'mf=20G'  # memory req,request the given resources
+            sgeJob.args.p = '0'      # Set priority to 0
+            # pecifies a range of priorities from -1023 to 1024.
+            # The higher the number, the higher the priority.
+            # The default priority for jobs is zero
+            sgeJob.args.command = '"' + self.deploy.cmd + '"'
+            sgeJob.args.b = 'y'      # This is a binary file
+            sgeJob.args.o = self.deploy.get_log_path() + '.sge'
+            cmd = str(sgeJob.execute(mode = 'echo'))
```

Comment:
It seems the scope of the "try" above can be reduced to a "with" for the file in line 54, and a separate try for the subprocess.

Reply:
fixed